dataplicity/remote_directory.py

```diff
@@ -0,0 +1,258 @@
+
+    def add_upload(self, path, upload_id):
+        # type: (Text, Text) -> int
+        """Add a new upload.
+
+        Args:
+            path (str): Path to the file to upload relative to self.path
+            upload_id (str): Upload ID.
+
+        Raises:
+            LowDiskSpace: If there is not enough disk space to create a snapshot (including margin)
+            AddFail: If the upload could not be added for any reason.
+
+        Returns:
+            int: The size of the snapshot file (in bytes)
+        """
+        file_path = os.path.join(self.path, path.lstrip("/"))
+
+        if not validate_path(file_path):
+            raise IllegalPath("Path %s is illegal" % file_path)
+
+        try:
+            disk_usage = disk_tools.disk_usage(self.temp_path)
+            file_size = snapshot_size = os.path.getsize(file_path)
+        except Exception as error:
+            # Don't want to fail here, probably worth honoring the request
+            log.warning("failed to get disk usage information; %s", error)
+        else:
+            if disk_usage.used + file_size > disk_usage.total - FREE_SPACE_MARGIN:
+                raise LowDiskSpace(
+                    "not enough space on %s to create snapshot" % self.temp_path
+                )
+
+        snapshot_path = self.get_snapshot_path(upload_id)
+        # Sanity check for path
+        if not snapshot_path.startswith(self.temp_path):
+            raise IllegalPath("Path %s is illegal" % path)
+
+        try:
+            # Copy file to temporary location, creating a "snapshot"
+            shutil.copyfile(file_path, snapshot_path)
+            snapshot_size = os.path.getsize(snapshot_path)
+            log.debug("created snapshot for %s; %s", file_path, snapshot_path)
+            return snapshot_size
+        except Exception as error:
+            log.error("failed to add_upload; %r", error)
+            raise AddFail("Unable to add upload; %s" % error)
+
+    def read_upload(self, upload_id, offset, size):
+        # type: (Text, int, int) -> bytes
+        """Read data from the snapshot.
+
+        Args:
+            upload_id (str): Upload ID.
+            offset (int): Offset in the file.
+            size (int): Number of bytes to read.
+
+        Raises:
+            ReadFail: If the file could not be read from.
+
+        Returns:
+            bytes: Raw data from the file.
+        """
+        snapshot_path = self.get_snapshot_path(upload_id)
+        try:
+            with open(snapshot_path, "rb") as snapshot_file:
+                snapshot_file.seek(offset)
+                data = snapshot_file.read(size)
+                log.debug("read %i bytes from %s", len(data), snapshot_path)
+        except Exception as error:
+            # Most likely, the snapshot file has been removed
+            log.error("failed to read_upload; %r", error)
+            raise ReadFail("unable to read %r" % upload_id)
+        return data
```

Comment:
the `return data` could as well be in line 167/168, right ?

Reply:
Could do. I like to put the return at the end wherever possible, so you don't need to go looking for it.